src/clients/echo/arcecho.1.in
.\" -*- nroff -*-
.TH ARCECHO 1 "@DATE@" "NorduGrid ARC @VERSION@" "NorduGrid Users Manual"

.SH NAME

arcecho \- ARC echo client

.SH DESCRIPTION

The
.B arcecho
command is a client for the ARC echo service.

.SH SYNOPSIS

.B arcecho [options] service message

.SH OPTIONS

.IP "\fB-z\fR, \fB--conffile\fR=\fIfilename\fR"
configuration file (default ~/.arc/client.conf)
.IP "\fB-d\fR, \fB--debug\fR=\fIdebuglevel\fR"
FATAL, ERROR, WARNING, INFO, VERBOSE or DEBUG
.IP "\fB-v\fR, \fB--version\fR"
print version information
.IP "\fB-?\fR, \fB--help\fR"
print help
.LP

.SH ARGUMENTS

.IP "\fBservice\fR"
A URL to an ARC echo service
.IP "\fBmessage\fR"
The message the service should return
.LP

.SH EXTENDED DESCRIPTION

.B arcecho
is a simple test tool: we can send a message to the echo service and
it will respond with the same message.

.SH FILES

.TP
.B ~/.arc/client.conf
Some options can be given default values by specifying them in the
ARC client configuration file. By using the
.B --conffile
option a different configuration file can be used than the default.

.SH ENVIRONMENT VARIABLES

.TP
.B X509_USER_PROXY
The location of the user's Grid proxy file. Shouldn't be set unless
the proxy is in a non-standard location.

.TP
.B ARC_LOCATION
The location where ARC is installed can be specified by this variable. If not
specified the install location will be determined from the path to the
command being executed, and if this fails a WARNING will be given stating the
location which will be used.

.TP
.B ARC_PLUGIN_PATH
The location of ARC plugins can be specified by this variable. Multiple
locations can be specified by separating them by : (; in Windows). The
default location is \fB$ARC_LOCATION\fR/lib/arc (\\ in Windows).

.SH COPYRIGHT

APACHE LICENSE Version 2.0

.SH AUTHOR

ARC software is developed by the NorduGrid Collaboration 
(http://www.nordugrid.org), please consult the AUTHORS file distributed with 
ARC. Please report bugs and feature requests to http://bugzilla.nordugrid.org



NEWS
no news good news :)


src/clients/echo/arcecho.cpp
// -*- indent-tabs-mode: nil -*-

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <iostream>
#include <list>
#include <string>

#include <arc/ArcLocation.h>
#include <arc/IString.h>
#include <arc/Logger.h>
#include <arc/OptionParser.h>
#include <arc/StringConv.h>
#include <arc/UserConfig.h>
#include <arc/communication/ClientInterface.h>
#include <arc/message/MCC.h>

int main(int argc, char **argv) {

  setlocale(LC_ALL, "");

  Arc::Logger logger(Arc::Logger::getRootLogger(), "arcecho");
  Arc::LogStream logcerr(std::cerr);
  logcerr.setFormat(Arc::ShortFormat);
  Arc::Logger::getRootLogger().addDestination(logcerr);
  Arc::Logger::rootLogger.setThreshold(Arc::WARNING);

  Arc::ArcLocation::Init(argv[0]);

  Arc::OptionParser options(istring("service message"),
                            istring("The arcecho command is a client for "
                                    "the ARC echo service."),
                            istring("The service argument is a URL to an ARC "
                                    "echo service.\n"
                                    "The message argument is the message the "
                                    "service should return."));

  int timeout = -1;
  options.AddOption('t', "timeout", istring("timeout in seconds (default 20)"),
                    istring("seconds"), timeout);

  std::string conffile;
  options.AddOption('z', "conffile",
                    istring("configuration file (default ~/.arc/client.conf)"),
                    istring("filename"), conffile);

  std::string debug;
  options.AddOption('d', "debug",
                    istring("FATAL, ERROR, WARNING, INFO, VERBOSE or DEBUG"),
                    istring("debuglevel"), debug);

  bool version = false;
  options.AddOption('v', "version", istring("print version information"),
                    version);

  std::list<std::string> args = options.Parse(argc, argv);

  if (version) {
    std::cout << Arc::IString("%s version %s", "arcecho", VERSION)
              << std::endl;
    return 0;
  }

  // If debug is specified as argument, it should be set before loading the configuration.
  if (!debug.empty())
    Arc::Logger::getRootLogger().setThreshold(Arc::string_to_level(debug));

  Arc::UserConfig usercfg(conffile);
  if (!usercfg) {
    logger.msg(Arc::ERROR, "Failed configuration initialization");
    return 1;
  }

  if (debug.empty() && !usercfg.Verbosity().empty())
    Arc::Logger::getRootLogger().setThreshold(Arc::string_to_level(usercfg.Verbosity()));

  if (timeout > 0) {
    usercfg.Timeout(timeout);
  }

  if (args.size() != 2) {
    logger.msg(Arc::ERROR, "Wrong number of arguments!");
    return 1;
  }

  std::list<std::string>::iterator it = args.begin();

  Arc::URL service = *it++;
  std::string message = *it++;

  Arc::MCCConfig cfg;
  usercfg.ApplyToConfig(cfg);

  Arc::ClientSOAP client(cfg, service, usercfg.Timeout());

  std::string xml;

  Arc::NS ns("echo", "http://www.nordugrid.org/schemas/echo");
  Arc::PayloadSOAP request(ns);
  request.NewChild("echo:echo").NewChild("echo:say") = message;

  request.GetXML(xml, true);
  logger.msg(Arc::INFO, "Request:\n%s", xml);

  Arc::PayloadSOAP *response = NULL;

  Arc::MCC_Status status = client.process(&request, &response);

  if (!status) {
    logger.msg(Arc::ERROR, (std::string)status);
    if (response)
      delete response;
    return 1;
  }

  if (!response) {
    logger.msg(Arc::ERROR, "No SOAP response");
    return 1;
  }

  response->GetXML(xml, true);
  logger.msg(Arc::INFO, "Response:\n%s", xml);
  if (response->IsFault()) {
    Arc::SOAPFault& fault = *response->Fault();
    std::string err("SOAP fault: %s", fault.Code());
    for (int n = 0;;++n) {
      if (fault.Reason(n).empty()) break;
      err += ": " + fault.Reason(n);
    }
    logger.msg(Arc::ERROR, err);
    return 1;
  }

  std::string answer = (std::string)((*response)["echoResponse"]["hear"]);

  delete response;

  std::cout << answer << std::endl;

  return 0;
}
